scli-runner/src/main/scala/com.olegych.scastie.sclirunner/BspClient.scala

```diff
@@ -146,33 +147,34 @@ class BspClient(private val workingDir: Path,
         .map(r => r.getItems())
         .map(classes => {
           if (classes.size() == 0) throw NoMainClassFound("No main class found.")
-          classes.get(0)
+          if (classes.get(0).getClasses().size() == 0) throw NoMainClassFound("No main class found.")
+          classes.get(0).getClasses().get(0)
         })
         // Run
         .flatMap(mainClass => {
-          wrapTimeout(s"$id-run", bspServer.buildTargetRun({
-            val param = new RunParams(targetId)
-            param.setDataKind("scala-main-class")
-            param.setData(mainClass.getClasses().get(0))
-            param.setOriginId(s"$id-run")
-            param
-          }))
+          bspServer.jvmRunEnvironment(new JvmRunEnvironmentParams(java.util.List.of(targetId)))
+            .asScala.map((mainClass, _))
         })
-        .map(result => {
-          if (result.getStatusCode() == StatusCode.ERROR) throw FailedRunError(s"Failed run: $result.")
-          if (result.getStatusCode() == StatusCode.CANCELLED) throw FailedRunError(s"Cancelled run: $result")
-          result
+        .map({
+          case (mainClass, runSettings) if runSettings.getItems().isEmpty() => throw BspClient.FailedRunError("Could not find run settings.")
+          case (mainClass, runSettings) => (mainClass, runSettings.getItems().get(0))
         })
-        .map(_ => {
-          val output = logMessages(s"$id-run")
-          logMessages.remove(s"$id-run")
-          runMessageEvent.remove(id)
+        .map({ case (mainClass, runSettings) => {
 
-          log.info(s"Ran successfully: $output")
-          BspClientRun(output.map(_.getMessage()))
-        }).recover { e =>
+          val classpath = runSettings.getClasspath.asScala.map(uri => Paths.get(new URI(uri))).mkString(":")
+          val envVars = Map(
+            "CLASSPATH" -> classpath
+          ) ++ runSettings.getEnvironmentVariables.asScala
+
+          val process = Process(
+            Seq("java", mainClass.getClassName()) ++ runSettings.getJvmOptions().asScala,
+            cwd = new java.io.File(runSettings.getWorkingDirectory()),
+            envVars.toSeq : _*
+          )
+
+          BspClientRun(process)
+        }}).recover { e =>
           logMessages.remove(s"$id-run")
-          runMessageEvent.remove(id)
           throw e
         }
       })
```

Comment:
Overall, as I told you before in the future ( not scala one, future work ), try to refactor it to separate calls so we can have final build methods constructed with calls like:

```scala
def build(blabla):
  initialize(blabla)
    .flatMap(reload(blablab))
    .flatMap(getMainClass(blabla))
    .flatMap(etc etc)
```

You can even do for comprehension then.